### PR TITLE
Exposes generation of diffs from base reporter

### PR DESCRIPTION
**Pull request by harrysarson**
#### Second attempt at solution for #3011, improves on PR #3201.

PR based on suggestion by @segrey in https://github.com/mochajs/mocha/pull/3201#issuecomment-358933824. Exposed function `generateDiff(expected, actual)` as a static member of `Base`.

The value of `Base.inlineDiffs` is used to determine the form of the diff.

### Alternate Designs

#3201

### Why should this be in core?

Allows IDE's to produce nice output when integrated with mocha.

### Benefits

See #3011.

### CODE REVIEW

**Comment by segrey** [comment]
Looks good to me. @HarrySarson thanks!

**Comment by segrey** [comment]
I'll update IntelliJ reporter once the PR is merged.

File: lib/reporters/base.js [hunk 1]
```diff
@@ -166,6 +166,25 @@ function stringifyDiffObjs (err) {
...
+ *
+ * @api private
+ * @param {String} actual
+ * @param {String} expected
+ * @return {string} Diff
+ */
+var generateDiff = exports.generateDiff = function (actual, expected) {
+  if (exports.inlineDiffs) {
+    return inlineDiff(actual, expected);
+  } else {
```
Comment (boneskull) [inline lib/reporters/base.js]: this `else` is not necessary

File: lib/reporters/base.js [hunk 2]
```diff
@@ -166,6 +166,25 @@ function stringifyDiffObjs (err) {
   }
 }
 
+/**
+ * Returns a diff between 2 strings with coloured ANSI output.
+ *
+ * The diff will be either inline or unified dependant on the value
+ * of `Base.inlineDiff`.
+ *
+ * @api private
```
Comment (boneskull) [inline lib/reporters/base.js]: I think the intent is that this is a public function, so you can remove this, right?

File: lib/reporters/base.js [hunk 3]
```diff
@@ -166,6 +166,25 @@ function stringifyDiffObjs (err) {
...
+ * @param {String} actual
```
Comment (boneskull) [inline lib/reporters/base.js]: lowercase `string`

File: test/reporters/base.spec.js [hunk 4]
```diff
@@ -136,6 +136,36 @@ describe('Base reporter', function () {
...
 
+  describe('Diff generation', function () {
+    it('should generate unified diffs if `inlineDiff === false`', function () {
+      var actual = 'a foo unified diff';
+      var expected = 'a bar unified diff';
+
+      var oldInlineDiffs = Base.inlineDiffs;
+      Base.inlineDiffs = false;
+
+      var output = Base.generateDiff(actual, expected);
```
Comment (boneskull) [inline test/reporters/base.spec.js]: please do the patching in a `beforeEach` of the suite and restore in `afterEach`

**Comment by boneskull** [comment]
@HarrySarson Thanks for this!! in the future, please just reuse your PRs and force-push changes--it's kind of difficult to track the conversation otherwise

**Comment by segrey** [comment]
@boneskull Regarding `err.details`, I meant to enhance standard `err` object that is passed to reporters in `runner.on('fail', function (test, err) {})`. The idea is to have some common properties that are calculated by mocha already and might be interested for reporters:
* `message` and `stack`: calculated using logic similar to https://github.com/mochajs/mocha/blob/v5.0.1/lib/reporters/base.js#L185-L205
* `diff`: the result of `generateDiff(err.actual, err.expected)`.

As a benefit, mocha will have a mechanism to easily change how these values are calculated, e.g.  supply diffs provided by assertion libraries without changes in reporters.

File: test/reporters/base.spec.js (hunk 4 above)
Comment (HarrySarson) [reply test/reporters/base.spec.js]: When doing the patching, should I be setting the value of `Base.inlineDiffs` in the `beforeEach`?

**Comment by HarrySarson** [comment]
Test failure is due to linting of markdown. Would be resolved by #3236

**Comment by harrysarson** [comment]
@boneskull can this be merged?

**Comment by boneskull** [comment]
@segrey Got it.  So it looks like this PR is the first step to get there.

**Comment by boneskull** [comment]
blah, meant to squash.

**Pull request merged**